### PR TITLE
make hostdb.round_robin_max_count configurable

### PR DESCRIPTION
**Pull request by zizhong**
@SolidWallOfCode @PSUdaemon could you please review this one?

### CODE REVIEW

File: iocore/hostdb/P_HostDBProcessor.h [hunk 1]
```diff
@@ -344,11 +348,10 @@ HostDBRoundRobin::select_best_srv(char *target, InkRand *rand, ink_time_t now, i
   }
 #endif
 
-  int i = 0, len = 0;
+  int i           = 0;
   uint32_t weight = 0, p = INT32_MAX;
   HostDBInfo *result = nullptr;
-  HostDBInfo *infos[HOST_DB_MAX_ROUND_ROBIN_INFO];
-
+  std::vector<HostDBInfo *> infos;
```
Comment (zizhong) [inline iocore/hostdb/P_HostDBProcessor.h]: Is `std::vector<HostDBInfo> infos(good)` okay?

File: iocore/dns/DNS.cc [hunk 2]
```diff
@@ -1575,30 +1573,31 @@ dns_process(DNSHandler *handler, HostEnt *buf, int len)
...
 
-        if (srv->host[0] != '\0') {
-          buf->srv_hosts.srv_hosts_length += srv->host_len;
+        if (srv.host[0] != '\0') {
+          buf->srv_hosts.srv_hosts_length += srv.host_len;
         } else {
           continue;
         }
-        ++num_srv;
+        buf->srv_hosts.hosts.emplace_back(srv);
```
Comment (zizhong) [inline iocore/dns/DNS.cc]: Right. I'll change it to `push_back`.

File: iocore/hostdb/P_HostDBProcessor.h (hunk 1 above)
Comment (SolidWallOfCode) [inline iocore/hostdb/P_HostDBProcessor.h]: No, that still does a heap allocation.

**Comment by SolidWallOfCode** [comment]
[approve ci]

File: iocore/hostdb/P_HostDBProcessor.h (hunk 1 above)
Comment (zizhong) [inline iocore/hostdb/P_HostDBProcessor.h]: Oh, you are saying VLA instead of new. I see.

**Comment by zizhong** [comment]
@SolidWallOfCode updated. Thanks for the suggestion.

**Comment by zizhong** [comment]
@SolidWallOfCode pushed another commit related to round_robin_max_count. Could you review it?

**Comment by bryancall** [comment]
[approve ci]

**Comment by zizhong** [comment]
@SolidWallOfCode ping. Looking forward to feedbacks from an ATS expert like you.

File: iocore/dns/SRV.h [hunk 3]
```diff
@@ -50,9 +49,8 @@ operator<(const SRV &left, const SRV &right)
 }
 
 struct SRVHosts {
-  unsigned int srv_host_count   = 0;
   unsigned int srv_hosts_length = 0;
-  SRV hosts[HOST_DB_MAX_ROUND_ROBIN_INFO];
+  std::vector<SRV> hosts;
```
Comment (a-canary) [inline iocore/dns/SRV.h]: I'm not sure if std::vector is the best structure for this. But I'd say go ahead and get this in before I start refactoring the HostDB code over the next few months.

Comment (zizhong) [inline iocore/dns/SRV.h]: Thanks! Glad to hear that.

**Comment by bryancall** [comment]
[approve ci]

File: iocore/dns/SRV.h (hunk 3 above)
Comment (zwoop) [inline iocore/dns/SRV.h]: It feels like struct SRVHosts is no longer needed ? Can we make this into a using (typedef) over the std::vector ?

Comment (zizhong) [inline iocore/dns/SRV.h]: If we remove `srv_hosts_length`, we need to loop over every time we need it. Do we really want to do that?

Comment (zwoop) [inline iocore/dns/SRV.h]: I see, the srv_host_length is the string length of all hosts. If so, I guess that's an ok optimization, I'd probably examine how often it's used, and if so, how expensive it is to iterator over it to get the lengths.

But if you decide to leave it, I'm fine with that.

Comment (bryancall) [inline iocore/dns/SRV.h]: That is a reasonable requirement in keeping `srv_hosts_length` around.

**Comment by bryancall** [comment]
@zizhong There are still build issues for this PR

File: iocore/dns/SRV.h (hunk 3 above)
Comment (zizhong) [inline iocore/dns/SRV.h]: Okay then. I'll leave it.

**Comment by zizhong** [comment]
@bryancall That's weird. The building is okay on my local. The `Details` link is broken now. Can you [approve ci] again?

**Comment by scw00** [comment]
[approve ci]

**Comment by zizhong** [comment]
Thanks, @scw00 . Can you please trigger the build again?

**Comment by scw00** [comment]
[approve ci]

**Comment by zizhong** [comment]
format fixed.

**Comment by bryancall** [comment]
[approve ci]

File: iocore/hostdb/HostDB.cc [hunk 4]
```diff
@@ -1382,8 +1384,8 @@ HostDBContinuation::dnsEvent(int event, HostEnt *e)
       if (is_srv()) {
         int skip  = 0;
         char *pos = (char *)rr_data + sizeof(HostDBRoundRobin) + valid_records * sizeof(HostDBInfo);
-        SRV *q[HOST_DB_MAX_ROUND_ROBIN_INFO];
-        ink_assert(valid_records <= HOST_DB_MAX_ROUND_ROBIN_INFO);
+        std::vector<SRV *> q(valid_records);
```
Comment (SolidWallOfCode) [inline iocore/hostdb/HostDB.cc]: Elsewhere you have
```
HostDBInfo *infos[good]
```

Why not something similar here, as the size of `valid_records` is know? That would avoid allocating memory for the `std::vector`.

**Comment by zizhong** [comment]
@SolidWallOfCode Fixed.

**Comment by scw00** [comment]
[approve ci]

**Comment by bryancall** [comment]
[approve ci]

**Pull request merged**